DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c

```diff
@@ -797,3 +835,198 @@ MergeFirmwareImages (
   FreePool (Data.Updated);
   return NULL;
 }
+
+/**
+  Checks if an image has Boot Guard enabled, by checking is KM and BPM are
+  present in the image.
+
+  @param[in] Image     Image to check
+  @param[in] ImageLen  Size of the image to check
+
+  @return TRUE    If Boot Guard BPM and KM are present
+  @return FALSE   If Boot Guard BPM and KM are absent
+**/
+STATIC
+BOOLEAN
+IsBootGuardEnabled(
+  IN CONST VOID  *Image,
+  IN CONST UINTN ImageLen
+  )
+{
+  struct cbfs_file   *File;
+  struct cbfs_image  Cbfs;
+  CONST Fmap         *FlashMap;
+
+  if (!GetFmap (Image, ImageLen, &FlashMap)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a(): failed to parse firmware\n",
+      __FUNCTION__
+      ));
+    return FALSE;
+  }
+
+  if (!GetCbfs (Image, FlashMap, &Cbfs)) {
+    DEBUG ((DEBUG_ERROR, "%a(): failed to load CBFS\n", __FUNCTION__));
+    return FALSE;
+  }
+
+  File = cbfs_get_entry (&Cbfs, "key_manifest.bin");
+  if (File == NULL)
+    return FALSE;
+
+  File = cbfs_get_entry (&Cbfs, "boot_policy_manifest.bin");
+  if (File == NULL)
+    return FALSE;
+
+  return TRUE;
+}
+
+/**
+  Extracts the BtgPubKey struct from a Key Manifest buffer.
+
+  @param[in]  Km             Key Manifest to parse
+  @param[in]  ImageLen       Size of the Key Manifest to check
+  @param[out] OemRootKey     The extracted OemRootKey
+  @param[out] OemRootKeySize Size of the extracted OemRootKey
+
+  @return EFI_INVALID_PARAMETER    If the input parameters or values in the Key
+                                   Manifest are invalid
+  @return EFI_SUCCESS              If extraction was successful
+**/
+STATIC
+EFI_STATUS
+GetOemRootKeyFromKm (
+  IN CONST VOID   *KmBuffer,
+  IN CONST UINTN  KmSize,
+  OUT VOID        **OemRootKey,
+  OUT UINTN       *OemRootKeySize
+  )
+{
+  BtgPubKey *PubKeyOffset, *PubKey;
+  KeyManifestHeader *Km = (KeyManifestHeader*) KmBuffer;
+
+  if (Km == NULL ||
+      OemRootKeySize == NULL ||
+      KmSize < sizeof(KeyManifestHeader) ||
+      KmSize < Km->KeySignatureOffset + sizeof(KeyAndSigHeader) + sizeof(BtgPubKey))
+    return EFI_INVALID_PARAMETER;
+
+  PubKeyOffset = (BtgPubKey*)((UINT8*)Km + Km->KeySignatureOffset + sizeof(KeyAndSigHeader));
+
+  if (PubKeyOffset->KeySize != 2048 && PubKeyOffset->KeySize != 3072)
+    return EFI_INVALID_PARAMETER;
+
+  if (PubKeyOffset->Exponent != 0x10001)
+    return EFI_INVALID_PARAMETER;
+
+  *OemRootKeySize = sizeof(BtgPubKey) + PubKeyOffset->KeySize / 8;
+
+  PubKey = AllocatePool (*OemRootKeySize);
+  CopyMem(PubKey, PubKeyOffset, *OemRootKeySize);
```

Comment:
There is a convenient shortcut for making a copy: `PubKey = AllocateCopyPool(PubKeyOffset, *OemRootKeySize);`.  However, why use dynamic allocation at all?  You're only reading the data inside of images while they exist.  As is, the code leaks the memory allocated here.